test/e2e/deploy_filesystem_test.go

```diff
@@ -0,0 +1,161 @@
+	appName := "test-deploy-filesystem"
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", appName})
+	}
+
+	cleanUp()
+	t.Cleanup(cleanUp)
+
+	theUI := ui.NewConfUI(ui.NewNoopLogger())
+	theUI.EnableNonInteractive()
+
+	configFactory := core.NewConfigFactoryImpl()
+	configFactory.ConfigureClient(100, 200)
+
+	// We don't need to customize any of these, but we'll get panics if they're nil
+	configFactory.ConfigurePathResolver(func() (string, error) {
+		return "", nil
+	})
+	configFactory.ConfigureContextResolver(func() (string, error) {
+		return "", nil
+	})
+	configFactory.ConfigureYAMLResolver(func() (string, error) {
+		return "", nil
+	})
+
+	depsFactory := core.NewDepsFactoryImpl(configFactory, theUI)
+	log := logger.NewUILogger(theUI)
+
+	deployOptions := app.NewDeployOptions(theUI, depsFactory, log)
+	deployOptions.AppFlags.NamespaceFlags.Name = env.Namespace
+	deployOptions.AppFlags.Name = appName
+	deployOptions.FileFlags.Files = []string{
+		"dir1",            // directory
+		"file1.yaml",      // file in root of fs
+		"dir2/file2.yaml", // file in subdir
+		"dir3",            // another directory
+	}
+
+	flagsFactory := core.NewFlagsFactory(configFactory, depsFactory)
+
+	deployCmd := app.NewDeployCmd(deployOptions, flagsFactory)
+
+	now := time.Now().Unix()
+	labelSelector := fmt.Sprintf("now=%d", now)
+
+	inMemFS := newFSBuilder().
+		file(
+			"dir1/cm1.yaml",
+			testConfigMap(env.Namespace, "cm1", now),
+		).
+		file(
+			"dir1/cm2.yaml",
+			testConfigMap(env.Namespace, "cm2", now),
+		).
+		file(
+			"file1.yaml",
+			testConfigMap(env.Namespace, "cm3", now),
+		).
+		file(
+			"dir2/file2.yaml",
+			testConfigMap(env.Namespace, "cm4", now),
+		).
+		file(
+			"dir3/cm5.yaml",
+			testConfigMap(env.Namespace, "cm5", now),
+		).
+		file(
+			"dir3/cm6.yaml",
+			testConfigMap(env.Namespace, "cm6", now),
+		).
+		toFS()
+
+	deployOptions.FileSystem = inMemFS
+
+	err := deployCmd.Execute()
+	require.NoError(t, err)
+
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	configOverrides := &clientcmd.ConfigOverrides{}
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, configOverrides)
+
+	restConfig, err := kubeConfig.ClientConfig()
+	require.NoError(t, err, "error creating rest config")
+
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	require.NoError(t, err, "error creating k8s clientset")
+
+	ctx := context.Background()
+	const allNamespaces = ""
+	configMaps, err := kubeClient.CoreV1().ConfigMaps(allNamespaces).List(ctx, metav1.ListOptions{LabelSelector: labelSelector})
+	require.NoError(t, err, "error listing ConfigMaps")
+	require.Len(t, configMaps.Items, 6)
+	expectedNames := sets.New[string]("cm1", "cm2", "cm3", "cm4", "cm5", "cm6")
+	actualNames := sets.New[string]()
+	for _, cm := range configMaps.Items {
+		actualNames.Insert(cm.Name)
+	}
+	require.Empty(t, cmp.Diff(expectedNames, actualNames))
```

Comment:
**nit**: Maybe we can use `require.ElementsMatch`? (We would still need to create an array of actualNames and expectedNames, but we won't have to use any other library)

Reply:
cmp.Diff is already in the repo, although it's only indirectly used before this change. Using cmp.Diff in tests provides extremely clean, easy to understand reasons why a test failed.

```go
m1 := []string{"a", "b"}
m2 := []string{"a", "c"}
assert.ElementsMatch(t, m1, m2)
assert.Empty(t, cmp.Diff(m1, m2))
```

The ElementsMatch error yields:
```
 Error:          elements differ
                                
                                extra elements in list A:
                                ([]interface {}) (len=1) {
                                 (string) (len=1) "b"
                                }
                                
                                
                                extra elements in list B:
                                ([]interface {}) (len=1) {
                                 (string) (len=1) "c"
                                }
                                
                                
                                listA:
                                ([]string) (len=2) {
                                 (string) (len=1) "a",
                                 (string) (len=1) "b"
                                }
                                
                                
                                listB:
                                ([]string) (len=2) {
                                 (string) (len=1) "a",
                                 (string) (len=1) "c"
                                }
```

The cmp.Diff error yields
```
Error:          Should be empty, but was   []string{
                                        "a",
                                -       "b",
                                +       "c",
                                  }

```